tcpreplay.8
.\" $Id: tcpreplay.8,v 1.12 2002/07/27 20:18:40 aturner Exp $
.TH TCPREPLAY 8
.SH NAME
tcpreplay \- replay packets from capture files
.SH SYNOPSIS
.na
.B tcpreplay
.B \-i
.I intf
[
.B \-c
.I cache file
] [
.B \-I
.I intf mac
] [
.B \-j
.I intf2
] [
.B \-J
.I intf mac
] [
.B \-l
.I loop count
] [
.B \-m
.I multiplier
|
.B \-r
.I rate
|
.B \-R
] [
.B \-C 
CIDR...
] [
.B \-u pad|trunc
] [
.B \-v
] [
.B \-h|-V
] [
.B \-M
]
.I file ...
.SH DESCRIPTION
.LP
.I tcpreplay
is a tool for replaying network traffic from files
saved with 
\fItcpdump(8)\fP
or
\fIsnoop(1M)\fP.
.LP
The basic operation of
.I tcpreplay
is to resend all packets from input file(s) at the speed at which
they were recorded, a specified data \fIrate\fP, or as fast as the hardware is 
capable.  
.LP
If no
\fIrate\fP or
\fImultiplier\fP are given, 
.I tcpreplay 
will replay packets at the speed they were captured from the wire.
.LP
By processing a \fIcache file\fP generated by \fItcpprep(8)\fP, tcpreplay 
is able to split traffic between two interfaces.  This is useful for testing 
bridges, routers, and other gateway devices. Also supported is rewriting of 
the destination \fIMAC addresses\fP to allow routers and other devices which 
proxy-arp to forward packets.
.SH OPTIONS
.LP
.TP
.B \-c
Specify the \fItcpprep cache\fR to use to process packets.
.TP
.B \-h
Prints help/usage
.TP
.B \-i
Specify the interface in which to send packets.
.TP
.B \-I
Specify the \fIdestination MAC\fR to use for packets being sent out the primary
interface.
.TP
.B \-j
Specify the \fIsecondary interface\fR in which to send packets.
.TP
.B \-J
Specify the \fIdestination MAC\fR to use for packets being sent out the 
secondary interface.
.TP
.B \-l
Resend the capture file(s) \fIloop count\fR times.  Setting this to 0 (zero)
will cause tcpreplay to loop infinately.
.TP
.B \-m
Resend the packets a \fImultiple\fR of the speed at which they were
recorded.
.TP
.B \-M
Disable sending martian packets (source networks: 0/8, 127/8, 255/8)
.TP
.B \-r
Resend the packets at the \fIrate\fR specified (in Mbps).
.TP
.B \-C
CIDR mode. Specify a list of CIDR's (network1/masklen1,network2/masklen2,...) 
to match against the source IP of each packet. Packets matching any of the 
CIDR's are sent out the primary interface; remaining packets are sent out 
the secondary interface.  Can't be used with -c.
.TP
.B \-R
Resend the packets as fast as possible. Negates other speed options.
.TP
.B \-u
Untruncates IP packets when snaplen > the actual packet length.  Packets can
be padded (\fIpad\fR) with zeros or truncated (\fItrunc\fR) to the snaplen.
Will also recalcuate the IP and TCP/UDP headers as necessary.
.TP
.B \-v
Verbose mode, currently unused.
.TP
.B \-V
Print version info and exit.
.SH "SEE ALSO"
tcpdump(8), snoop(1M), tcpprep(1), capinfo(1)
.SH AUTHORS
Matt Undy, Anzen Computing.
.br
Matt Bing <mbing@nfr.net>
.br
Aaron Turner <aturner@pobox.com>
.SH AVAILABILITY
.LP
The current version is available via HTTP:
.LP
.RS
.I http://www.sourceforge.net/projects/tcpreplay/
.RE
.SH BUGS
.I tcpreplay
can only send packets as fast as your machine's interface,
processor, and disk will allow.
.LP
``\fIN\fR write attempts failed from full buffers and were repeated''
indicates that kernel buffers were full and \fIN\fR packets had to
be resent.
.LP
Looping captured traffic may simulate odd conditions on a network.
For example, opening the same TCP connection multiple times may 
exhaust resources on machines tracking the connection. The 
.B \-l
flag is provided to allow faster sending on machines with greater I/O
constraints. 
.LP
Looping will not work when reading from stdin since tcpreplay closes 
then re-opens the file and re-reads it from the beginning.
.LP
On BSD-based systems, kernel modifications may be required to preserve
outgoing link layer source addresses. Refer to the \fIlibnet(3)\fP
documentation for more information on how to do this. Increasing
NMBCLUSTERS kernel option will help performance.


CHANGELOG
xx/xx/xxxx: Version 1.2
	- Added CIDR dual-nic support to tcpreplay. 
	  patch by: Scott Mace <smace@intt.org>

06/17/2002: Version 1.1
	- Major rewrite
	- Support multiple nics
	- Better control over packet rates
	- Added support for snoop capture files
	- Includes tcpprep and capinfo commands


tcpreplay.c
/* $Id: tcpreplay.c,v 1.39 2002/10/22 05:01:21 aturner Exp $ */

#include "config.h"

#include <ctype.h>
#include <err.h>
#include <fcntl.h>
#include <libnet.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#include "tcpreplay.h"
#include "cache.h"
#include "cidr.h"
#include "libpcap.h"
#include "snoop.h"
#include "list.h"
#include "err.h"
#include "do_packets.h"
#include "xX.h"

struct options options;
char *cachedata = NULL;
CIDR *cidrdata = NULL;
struct timeval begin, end;
unsigned long bytes_sent, failed, pkts_sent;
char *cache_file = NULL, *intf = NULL, *intf2 = NULL;
int cache_bit, cache_byte, cache_packets;
volatile int didsig;

int include_exclude_mode;
CIDR *xX_cidr = NULL;
LIST *xX_list = NULL;


#ifdef DEBUG
int debug = 0;
#endif

void replay_file(char *);


void packet_stats();
void usage();
void version();
void mac2hex(const char *, char *, int); 
void configfile(char *);
int argv_create(char *, int, char **);

int
main(int argc, char *argv[])
{
	char ebuf[256]; 
	int ch, i;
	void *xX;

	bytes_sent = failed = pkts_sent = 0;
	intf = intf2 = NULL;
	memset(&options, 0, sizeof(options));

	/* Default mode is to replay pcap once in real-time */
	options.mult = 1.0;
	options.n_iter = 1;
	options.rate = 0.0;

	cache_bit = cache_byte = 0;

#ifdef DEBUG
	while ((ch = getopt(argc, argv, "dc:C:f:hi:I:j:J:l:m:Mr:Rs:u:Vvx:X:?")) != -1)
#else
	while ((ch = getopt(argc, argv, "c:C:f:hi:I:j:J:l:m:Mr:Rs:u:Vvx:X:?")) != -1)
#endif
		switch(ch) {
		case 'c': /* cache file */
			cache_file = optarg;
			cache_packets = read_cache(&cachedata, cache_file);
			break;
		case 'C': /* cidr matching */
			options.cidr = 1;
			if (!parse_cidr(&cidrdata, optarg))
				usage();
			break;
#ifdef DEBUG
		case 'd': /* enable debug */
			debug = 1;
			break;
#endif
		case 'f': /* config file*/
			configfile(optarg);
			break;
		case 'i': /* interface */
			intf = optarg;
			break;
		case 'I': /* primary dest mac */
			mac2hex(optarg, options.intf1_mac, sizeof(options.intf1_mac));
			if (memcmp(options.intf1_mac, NULL_MAC, 6) == 0)
				errx(1, "Invalid mac address: %s", optarg);
			break;
		case 'j': /* secondary interface */
			intf2 = optarg;
			break;
		case 'J': /* secondary dest mac */
			mac2hex(optarg, options.intf2_mac, sizeof(options.intf2_mac));
			if (memcmp(options.intf2_mac, NULL_MAC, 6) == 0)
				errx(1, "Invalid mac address: %s", optarg);
			break;
		case 'l': /* loop count */
			options.n_iter = atoi(optarg);
			if (options.n_iter < 0)
				errx(1, "Invalid loop count: %s", optarg);
			break;
		case 'm': /* multiplier */
			options.mult = atof(optarg);
			if (options.mult <= 0)
				errx(1, "Invalid multiplier: %s", optarg);
			options.rate = 0.0;
			break;
		case 'M': /* disable sending martians */
			options.no_martians = 1;
			break;
		case 'r': /* target rate */
			options.rate = atof(optarg);			
			if (options.rate <= 0)
				errx(1, "Invalid rate: %s", optarg);
			/* convert to bytes */
			options.rate = (options.rate * (1024*1024)) / 8;
			options.mult = 0.0;
			break;
		case 'R': /* replay at top speed */
			options.topspeed = 1;
			break;
		case 's':
			options.seed = atoi(optarg);
			break;
		case 'v': /* verbose */
			options.verbose++;
			break;
		case 'u': /* untruncate packet */
			if (strcmp("pad", optarg) == 0) {
				options.trunc = PAD_PACKET;
			} else if (strcmp("trunc", optarg) == 0) {
				options.trunc = TRUNC_PACKET;
			} else {
				errx(1, "Invalid untruncate option: %s", optarg);
			}
			break;
		case 'V':
			version();
			break;
		case 'x':
			include_exclude_mode = optind;
			if ((xX = parse_xX_str(include_exclude_mode, optarg)) == NULL)
				errx(1, "Unable to parse -x: %s", optarg);
			if (include_exclude_mode & xXPacket) {
				xX_list = (LIST *)xX;
			} else {
				xX_cidr = (CIDR *)xX;
			}
			break;
		case 'X':
			include_exclude_mode = optind;
			if ((xX = parse_xX_str(include_exclude_mode, optarg)) == NULL)
				errx(1, "Unable to parse -X: %s", optarg);
			if (include_exclude_mode & xXPacket) {
				xX_list = (LIST *)xX;
			} else {
				xX_cidr = (CIDR *)xX;
			}
			break;
		default:
			usage();
		}

	argc -= optind;
	argv += optind;

	if ( (options.mult > 0.0 && options.rate > 0.0) || argc == 0)
		usage();

	if (argc > 1)
		for (i = 0; i < argc; i++)
			if (!strcmp("-", argv[i]))
				errx(1, "stdin must be the only file specified");

	if (intf == NULL)
		errx(1, "Must specify interface");

	if ((intf2 == NULL) && (cache_file != NULL))
		errx(1, "Needs secondary interface with cache");

	if ((intf2 != NULL) && (!options.cidr && (cache_file == NULL) ))
		errx(1, "Needs cache or cidr match with secondary interface");

	/* process -x and -X 
	 * need to fix this -ADT
	if ((options.include != NULL) && (options.exclude != NULL))
		errx(1, "Can only specify -x or -X, not both");
	*/
	/* use our seed to make pseudo-random IP's */
	if (options.seed != 0) {
		srand(options.seed);
		options.seed = random();
#ifdef DEBUG
		dbg(1, "random() picked: %d", options.seed);
#endif
	}
		

#if USE_LIBNET_VERSION == 10
	if ((options.intf1 = libnet_open_link_interface(intf, ebuf)) == NULL)
		errx(1, "Can't open %s: %s", intf, ebuf);

	if (options.intf1->device == NULL)
		options.intf1->device = intf;

	if (intf2 != NULL) { 
		if ((options.intf2 = libnet_open_link_interface(intf2, ebuf)) == NULL)
			errx(1, "Can't open %s: %s", intf2, ebuf);

		if (options.intf2->device == NULL)
			options.intf2->device = intf2;
	}

#elif USE_LIBNET_VERSION == 11
	if ((options.intf1 = libnet_init(LIBNET_LINK_ADV, intf, ebuf)) == NULL)
		errx(1, "Can't open %s: %s", intf, ebuf);

	if (intf2 != NULL) {
		if ((options.intf2 = libnet_init(LIBNET_LINK_ADV, intf2, ebuf)) == NULL)
			errx(1, "Can't open %s: %s", intf2, ebuf);
	}
#endif

	warnx("sending on %s %s", intf, intf2 == NULL ? "" : intf2);

	if (gettimeofday(&begin, NULL) < 0)
		err(1, "gettimeofday");

	/* main loop */
	if (options.n_iter > 0) { 
		while (options.n_iter--) { /* limited loop */
			for (i = 0; i < argc; i++) {
				/* reset cache markers for each iteration */
				cache_byte = 0;
				cache_bit = 0;
				replay_file(argv[i]);
			}
		}
	} else { /* loop forever */
		while (1) {
			for (i = 0; i < argc; i++) {
				/* reset cache markers for each iteration */
				cache_byte = 0;
				cache_bit = 0;
				replay_file(argv[i]);
			}
		}
	}

	if (bytes_sent > 0)
		packet_stats();

	return 0;
}



void
replay_file(char *path)
{
	int fd;

	if (!strcmp(path, "-")) {
		fd = STDIN_FILENO;
	} else if ((fd = open(path, O_RDONLY, 0)) < 0) {
		warn("skipping %s: could not open", path);
		return;
	}

	if (is_snoop(fd)) {
#ifdef DEBUG
		if (debug)
			warnx("File %s is a snoop file", path);
#endif
		do_packets(fd, get_next_snoop);
		(void)close(fd);
	} else if (is_pcap(fd)) {
#ifdef DEBUG
		if (debug)
			warnx("File %s is a pcap file", path);
#endif
		do_packets(fd, get_next_pcap);
		(void)close(fd);
	} else {
		warnx("skipping %s: unknown format", path);
	}
}


void
packet_stats()
{
	float bytes_sec = 0.0, mb_sec = 0.0;
	int pkts_sec = 0;

	if (gettimeofday(&end, NULL) < 0)
		err(1, "gettimeofday");

	timersub(&end, &begin, &begin);
	if (timerisset(&begin)) {
		if (bytes_sent) {
		bytes_sec = bytes_sent / (begin.tv_sec + (float)begin.tv_usec / 1000000);
		mb_sec = (bytes_sec * 8) / (1024 * 1024);
		}
		if (pkts_sent)
		pkts_sec = pkts_sent / (begin.tv_sec + (float)begin.tv_usec / 1000000);
	}

	fprintf(stderr, " %ld packets (%ld bytes) sent in %ld seconds\n",
		pkts_sent, bytes_sent, begin.tv_sec);
	fprintf(stderr, " %.1f bytes/sec %.2f megabits/sec %d packets/sec\n", 
		bytes_sec, mb_sec, pkts_sec);

	if (failed) {
		fprintf(stderr, 
			" %ld write attempts failed from full buffers and were repeated\n",
			failed);
	}
}

/*
 * converts a string representation of a MAC address, based on 
 * non-portable ether_aton() 
 */
void
mac2hex(const char *mac, char *dst, int len) 
{
	int i;
	long l;
	char *pp;

	if (len < 6)
		return;

	while (isspace(*mac))
		mac++;

	/* expect 6 hex octets separated by ':' or space/NUL if last octet */ 
	for (i = 0; i < 6; i++) { 
		l = strtol(mac, &pp, 16); 
		if (pp == mac || l > 0xFF || l < 0) 
			return; 
		if (!(*pp == ':' || (i == 5 && (isspace(*pp) || *pp == '\0')))) 
			return; 
		dst[i] = (u_char)l;
		mac = pp + 1; 
	}
}

/* whorishly appropriated from fragroute-1.2 */
#define MAX_ARGS 128
int
argv_create(char *p, int argc, char *argv[])
{
	int i;
	
	for (i = 0; i < argc - 1; i++) {
		while (*p != '\0' && isspace((int)*p))
			*p++ = '\0';
		
		if (*p == '\0')
			break;
		argv[i] = p;
		
		while (*p != '\0' && !isspace((int)*p))
			p++;
	}
	p[0] = '\0';
	argv[i] = NULL;
	
	return (i);
}

void 
configfile(char *file) {
	FILE *fp;
	char *argv[MAX_ARGS], buf[BUFSIZ];
	int argc, i;

	if ((fp = fopen(file, "r")) == NULL)
		errx(1, "Could not open config file %s", file);

	for (i = 1; fgets(buf, sizeof(buf), fp) != NULL; i++) {
		if (*buf == '#' || *buf == '\r' || *buf == '\n')
			continue;

		if ((argc = argv_create(buf, MAX_ARGS, argv)) < 1) {
			warnx("couldn't parse arguments (line %d)", i);
			break;
		}

#define ARGS(x, y) ( (!strcmp(argv[0], x)) && (argc == y) )
		if (ARGS("cachefile", 2)) {
			cache_file = strdup(argv[1]);
			cache_packets = read_cache(&cachedata, cache_file);
		} else if (ARGS("cidr", 2)) {
			options.cidr = 1;
			if (!parse_cidr(&cidrdata, argv[1]))
				usage();
#ifdef DEBUG
		} else if (ARGS("debug", 1)) {
			debug = 1;
#endif
		} else if (ARGS("intf", 2)) {
			intf = strdup(argv[1]);
		} else if (ARGS("primary_mac", 2)) {
			mac2hex(argv[1], options.intf1_mac, sizeof(options.intf1_mac));
			if (memcmp(options.intf1_mac, NULL_MAC, 6) == 0)
				errx(1, "Invalid mac address: %s", argv[1]);
		} else if (ARGS("second_intf", 2)) {
			intf2 = strdup(argv[1]);
		} else if (ARGS("second_mac", 2)) {
			mac2hex(argv[1], options.intf2_mac, sizeof(options.intf2_mac));
			if (memcmp(options.intf2_mac, NULL_MAC, 6) == 0)
				errx(1, "Invalid mac address: %s", argv[1]);
		} else if (ARGS("loop", 2)) {
			options.n_iter = atoi(argv[1]);
			if (options.n_iter < 0)
				errx(1, "Invalid loop count: %s", argv[1]);
		} else if (ARGS("multiplier", 2)) {
			options.mult = atof(argv[1]);
			if (options.mult <= 0)
				errx(1, "Invalid multiplier: %s", argv[1]);
			options.rate = 0.0;
		} else if (ARGS("no_martians", 1)) {
			options.no_martians = 1;
		} else if (ARGS("rate", 2)) {
			options.rate = atof(argv[1]);			
			if (options.rate <= 0)
				errx(1, "Invalid rate: %s", argv[1]);
			/* convert to bytes */
			options.rate = (options.rate * (1024*1024)) / 8;
			options.mult = 0.0;
		} else if (ARGS("topspeed", 1)) {
			options.topspeed = 1;
		} else if (ARGS("verbose", 1)) {
			options.verbose++;
		} else if (ARGS("untruncate", 2)) {
			if (strcmp("pad", argv[1]) == 0) {
				options.trunc = PAD_PACKET;
			} else if (strcmp("trunc", argv[1]) == 0) {
				options.trunc = TRUNC_PACKET;
			} else {
				errx(1, "Invalid untruncate option: %s", argv[1]);
			}
		} else if (ARGS("seed", 2)) {
			options.seed = atol(argv[1]);
/*
 * need to fix this -ADT
		} else if (ARGS("include", 2)) {
			options.include = strdup(argv[1]);
		} else if (ARGS("exclude", 2)) {
			options.exclude = strdup(argv[1]);
*/
		} else {
			errx(1, "Skipping unrecognized: %s", argv[0]);
		}
	}
}

void
version()
{
	fprintf(stderr, "tcpreplay version: %s\n", VERSION);
	fprintf(stderr, "Compiled against libnet: %s\n", LIBNET_VERSION);
	exit(0);
}

void
usage()
{
	fprintf(stderr, "Usage: tcpreplay\n");
	fprintf(stderr, "-c <cachefile>\t\tSplit traffic via cache file\n"
			"-C CIDR1,CIDR2,...\tSplit traffic in CIDR Mode\n");
#ifdef DEBUG
	fprintf(stderr, "-d\t\t\tEnable debug output to STDERR\n");
#endif
	fprintf(stderr, "-f <configfile>\t\tSpecify configuration file\n"
			"-h\t\t\tHelp\n"
			"-i <nic>\t\tPrimary interface to send traffic out of\n"
			"-I <mac>\t\tRewrite dest MAC on primary interface\n"
			"-j <nic>\t\tSecondary interface to send traffic out of\n"
			"-J <mac>\t\tRewrite dest MAC on secondary interface\n"
			"-l <loop>\t\tSpecify number of times to loop\n"
			"-m <multiple>\t\tSet replay speed to given multiple\n"
			"-M\t\t\tDisable sending martian IP packets\n"
			"-r <rate>\t\tSet replay speed to given rate (Mbps)\n"
			"-R\t\t\tSet replay speed to as fast as possible\n"
			"-s <seed>\t\tRandomize src/dst IP addresses w/ given seed\n"
			"-u pad|trunc\t\tPad/Truncate packets which are larger than the snaplen\n"
			"-v\t\t\tVerbose\n"
			"-V\t\t\tVersion\n"
			"-x <match>\t\tOnly send the packets specified\n"
			"-X <match>\t\tSend all the packets except those specified\n"
			"<file1> <file2> ...\tFile list to replay\n");
	exit(1);
}
